my/core/query.py

```diff
@@ -0,0 +1,386 @@
+    pass
+
+
+def locate_function(module_name: str, function_name: str) -> Callable[[], Iterable[ET]]:
+    """
+    Given a module name and a function, returns the corresponding function.
+    Since we're in the query module, it is assumed that this returns an
+    iterable of objects of some kind, which we want to query over, though
+    that isn't required
+    """
+    try:
+        mod = importlib.import_module(module_name)
+        for (fname, func) in inspect.getmembers(mod, inspect.isfunction):
+            if fname == function_name:
+                return func
+    except Exception as e:
+        raise QueryException(str(e))
+    raise QueryException(f"Could not find function {function_name} in {module_name}")
+
+
+timedelta_regex = re.compile(r"^((?P<days>[\.\d]+?)d)?((?P<hours>[\.\d]+?)h)?((?P<minutes>[\.\d]+?)m)?((?P<seconds>[\.\d]+?)s)?$")
+
+
+# https://stackoverflow.com/a/51916936
+def parse_timedelta_string(timedelta_str: str) -> timedelta:
+    """
+    This uses a syntax similar to the 'GNU sleep' command
+    e.g.: 10d5h10m50s means '10 days, 5 hours, 10 minutes, 50 seconds'
+    """
+    parts = timedelta_regex.match(timedelta_str)
+    if parts is None:
+        raise ValueError(f"Could not parse time duration from {timedelta_str}.\nValid examples: '8h', '2d8h5m20s', '2m4s'")
+    time_params = {name: float(param) for name, param in parts.groupdict().items() if param}
+    return timedelta(**time_params)  # type: ignore[arg-type]
+
+
+
+def _generate_order_by_func(
+        obj_res: Res[T],
+        key: Optional[str] = None,
+        where_function: Optional[Where] = None,
+        default: Optional[U] = None
+) -> Optional[OrderFunc]:
+    """
+    Accepts an object Res[T] (Instance of some class or Exception)
+
+    If its an error, the generated function returns None
+
+    Most of the time, you'd want to provide at least a 'key', a 'where_function' or a 'default'.
+    You can provide both a 'where_function' and a default, or a 'key' and a default,
+    incase the 'where_function' doesn't work for a particular type/you hit an error
+
+    If a 'default' is provided, it is used for Exceptions and if an
+    OrderFunc function could not be determined for this type
+
+    If a key is given (the user specified which attribute), the function
+    returns that key from the object
+    tries to find that key on the object
+
+    Attempts to find an attribute which matches the 'where_function' on the object,
+    using some getattr/dict checks. Returns a function which when called with
+    this object returns the value to order by
+    """
+    if isinstance(obj_res, Exception):
+        if default is not None:
+            return lambda _o: default
+        else:
+            low(f"""While creating order_by function, encountered exception {obj_res}
+Value to order_by unknown, provide a 'default', filter exceptons with a 'where' predicate or
+pass 'drop_errors' to ignore this""")
+            return lambda _o: None
```

Comment:
This would probably raise when actually trying to compare? Perhaps could use some sort of special 'terminal object' which always compares `False` or `True` (depending on whether we want these items to show in the beginning or in the end). This might also make `default` argument from the `select` interface unnecessary? Although don't mind having it either -- guess just don't know how useful it would be in advance

Imo, even better, Exceptions and objects for which the predicate fails to compute in general would stick to the closest 'good' entry -- that way might be easier to debug. 
Got an experimental helper for that but it's not very generic, and either way can improve this later
 https://github.com/karlicoss/HPI/blob/eb26cf863345f764a8eb279aff0589d67e397bae/my/core/error.py#L53-L57

Reply:
Yeah, It would probably fail with a 'cant compare NoneType' error? 

I was originally trying some way to push exceptions and None to the front of the list, as they're sort of 'unsortable', but I couldn't figure out what the 'terminal object' should be, since for example:

If the user passes an `order_value` function and it trying to find datetimes `order_value=lambda o: isinstance(o, datetime)`, this would then be trying to compare `None` to a `datetime`, but theres no way for this to know that without an additional `O(n)` check (maybe thats fine though)

At one point I had another flag to `drop_unsorted`, which removed error/NoneType before, but I thought it might be a bit too expensive and its essentially another linear computation. Perhaps that doesn't really matter though -- the ability to sort by mixed types means that its worth it.

I think perhaps its possible to add another linear check [here](https://github.com/seanbreckenridge/HPI-to-master/blob/query/my/core/query.py#L359) that calls the `order_by_chosen` functions on each item in the iterable. If any of those return None it could:

- if `drop_unsorted` is present, drop them
- else, put unsorted items at the end of the list

Reply:
by a terminal object I meant something like this:

```
[ins] In [20]: @total_ordering 
          ...: class Dummy: 
          ...:     def __le__(self, other): 
          ...:         return True 
          ...:                                                                                                                                                                                                                                                

[ins] In [21]: dummy = Dummy()                                                                                                                                                                                                                                

[ins] In [22]: list(sorted([5,  None, 3, 9, None], key=lambda i: dummy if i is None else i))                                                                                                                                                                  
Out[22]: [None, None, 3, 5, 9]
```

So `dummy` now will always make the error entries stick to one side of the list.
I could swear something like this was in stdlib, but can't find it 🤔

Reply:
Ah, that makes sense -- would just wrap unsortable items into that.

I'll experiment with that. Should probably be enabled by default (just like how `Iterator[Res[Model]]` functions can return errors, this should also just forward errors back), with a flag to disable it (which could cause errors, but thats a feature you're enabling then...

will probably add both `wrap_unsorted` (True by default) and `drop_unsorted` (False by default)